extra_data/reader.py

```diff
@@ -708,7 +712,50 @@ def select(self, seln_or_source_glob, key_glob='*'):
         files = [f for f in self.files
                  if f.all_sources.intersection(selection.keys())]
 
-        return DataCollection(files, selection=selection, train_ids=self.train_ids)
+        if require_all:
+            # Select only those trains for which all selected sources
+            # and keys have data, i.e. have a count > 0 in their
+            # respective INDEX section.
+
+            train_ids = self.train_ids
+
+            for source, keys in selection.items():
+                if source in self.instrument_sources:
+                    # For INSTRUMENT sources, the INDEX is saved by
+                    # key group, which is the first hash component. In
+                    # practice, this is always 'data', but this is not
+                    # guaranteed and there may be more than one.
```

Comment:
```suggestion
                    # key group, which is the first hash component. In
                    # many cases this is 'data', but not always.
```

Modules of the 'big' detectors (AGIPD, LPD, DSSC) have 4 key groups: detector, header, image, trailer. I think I've been told at some point that the 'key groups' (as we call them) are designed around those detectors.